File: strategypattern/strategypattern.js
// ===== Strategy Interface (simulated with base class) =====
class DriveStrategy {
  drive() {
    throw new Error('Method "drive()" must be implemented');
  }
}

// ===== Concrete Strategies =====
class NormalDriveStrategy extends DriveStrategy {
  drive() {
    console.log("Normal Drive Strategy");
  }
}

class SportsDriveStrategy extends DriveStrategy {
  drive() {
    console.log("Sports Drive Strategy");
  }
}

// ===== Vehicle (Context) =====
class Vehicle {
  constructor(strategy) {
    this.strategy = strategy;
  }

  drive() {
    this.strategy.drive();
  }
}

// ===== Concrete Vehicles (Configured with Strategy) =====
class OffRoadVehicle extends Vehicle {
  constructor() {
    super(new SportsDriveStrategy());
  }
}

class PassengerVehicle extends Vehicle {
  constructor() {
    super(new NormalDriveStrategy());
  }
}

// ===== Test the implementation =====
new PassengerVehicle().drive();  // Output: Normal Drive Strategy
new OffRoadVehicle().drive();    // Output: Sports Drive Strategy

File: strategypattern/strategypattern.py
from abc import ABC, abstractmethod

class DriveStrategy(ABC):
    @abstractmethod
    def drive(self):
        pass

class NormalDriveStrategy(DriveStrategy):
    def drive(self):
        print('Normal Drive Strategy')

class SportsDriveStrategy(DriveStrategy):
    def drive(self):
        print('Sports Drive Strategy')

class Vehicle:
    def __init__(self, strategy):
        self.strategy = strategy

    def drive(self):
        self.strategy.drive()

class OffRoadVehicle(Vehicle):
    def __init__(self):
        super().__init__(SportsDriveStrategy())

class PassengerVehicle(Vehicle):
    def __init__(self):
        super().__init__(NormalDriveStrategy())

# Testing the implementation
PassengerVehicle().drive()  # Output: Normal Drive Strategy
OffRoadVehicle().drive()    # Output: Sports Drive Strategy
